File: aamva_parser.R
# =====================================================
# AAMVA Driver's License Barcode Parser
# License to Live: MIAS Project
# =====================================================

library(stringr)
library(lubridate)

#' Parse AAMVA Driver's License Barcode
#'
#' @param barcode_data Raw string from 2D barcode scanner
#' @return List with parsed fields
#' @export
parse_aamva_barcode <- function(barcode_data) {
  
  # Initialize result list
  result <- list(
    raw_data = barcode_data,
    parsed = FALSE,
    error = NULL
  )
  
  # Try to parse the barcode
  tryCatch({
    
    # CRITICAL: Clean up Eyoyo scanner output
    # 1. Remove leading @ symbol
    barcode_data <- gsub("^@\\s*", "", barcode_data)
    
    # 2. Remove ALL line breaks and newlines (scanner splits data across lines)
    barcode_data <- gsub("[\r\n]+", "", barcode_data)
    
    # 3. Remove any remaining whitespace at start/end
    barcode_data <- str_trim(barcode_data)
    
    # AAMVA Field Codes
    fields <- list(
      # Personal Information
      first_name = extract_field(barcode_data, "DAC"),
      middle_name = extract_field(barcode_data, "DAD"),
      last_name = extract_field(barcode_data, "DCS"),
      
      # License Information
      license_number = extract_field(barcode_data, "DAQ"),
      
      # Date Information
      date_of_birth = parse_aamva_date(extract_field(barcode_data, "DBB")),
      issue_date = parse_aamva_date(extract_field(barcode_data, "DBD")),
      expiration_date = parse_aamva_date(extract_field(barcode_data, "DBA")),
      
      # Address Information
      address_street = extract_field(barcode_data, "DAG"),
      address_city = extract_field(barcode_data, "DAI"),
      address_state = extract_field(barcode_data, "DAJ"),
      address_zip = extract_field(barcode_data, "DAK"),
      
      # Physical Description
      sex = decode_sex(extract_field(barcode_data, "DBC")),
      height_inches = extract_field(barcode_data, "DAU"),
      weight_lbs = extract_field(barcode_data, "DAW"),
      eye_color = decode_eye_color(extract_field(barcode_data, "DAY")),
      hair_color = decode_hair_color(extract_field(barcode_data, "DAZ")),
      
      # Additional Information
      country = extract_field(barcode_data, "DCG"),
      document_discriminator = extract_field(barcode_data, "DCF")
    )
    
    result$fields <- fields
    result$parsed <- TRUE
    
  }, error = function(e) {
    result$error <- paste("Parse error:", e$message)
    result$parsed <- FALSE
  })
  
  return(result)
}

#' Extract a specific field from AAMVA barcode data
#'
#' @param data Barcode string
#' @param code AAMVA field code (e.g., "DAC", "DCS")
#' @return Extracted value or NA
extract_field <- function(data, code) {
  # Pattern: code followed by data until next code or end
  pattern <- paste0(code, "([^D]*?)(?=D[A-Z]{2}|$)")
  match <- str_match(data, pattern)
  
  if (!is.na(match[1, 2])) {
    value <- str_trim(match[1, 2])
    # Remove trailing non-alphanumeric except spaces and hyphens
    value <- str_replace(value, "[^A-Za-z0-9 -]+$", "")
    return(value)
  }
  return(NA)
}

#' Parse AAMVA date format (MMDDYYYY)
#'
#' @param date_str Date string in MMDDYYYY format
#' @return Date object or NA
parse_aamva_date <- function(date_str) {
  if (is.na(date_str) || nchar(date_str) != 8) {
    return(NA)
  }
  
  tryCatch({
    month <- substr(date_str, 1, 2)
    day <- substr(date_str, 3, 4)
    year <- substr(date_str, 5, 8)
    date_string <- paste(year, month, day, sep = "-")
    return(as.Date(date_string))
  }, error = function(e) {
    return(NA)
  })
}

#' Decode sex code
#'
#' @param code Sex code (1 = Male, 2 = Female)
#' @return "Male", "Female", or NA
decode_sex <- function(code) {
  if (is.na(code)) return(NA)
  
  sex_map <- c("1" = "Male", "2" = "Female", "M" = "Male", "F" = "Female")
  return(sex_map[code])
}

#' Decode eye color code
#'
#' @param code Eye color abbreviation
#' @return Full eye color name
decode_eye_color <- function(code) {
  if (is.na(code)) return(NA)
  
  eye_colors <- c(
    "BLK" = "Black",
    "BLU" = "Blue",
    "BRO" = "Brown",
    "GRY" = "Gray",
    "GRN" = "Green",
    "HAZ" = "Hazel",
    "MAR" = "Maroon",
    "PNK" = "Pink",
    "DIC" = "Dichromatic"
  )
  
  result <- eye_colors[code]
  return(ifelse(is.na(result), code, result))
}

#' Decode hair color code
#'
#' @param code Hair color abbreviation
#' @return Full hair color name
decode_hair_color <- function(code) {
  if (is.na(code)) return(NA)
  
  hair_colors <- c(
    "BAL" = "Bald",
    "BLK" = "Black",
    "BLN" = "Blond",
    "BRO" = "Brown",
    "GRY" = "Gray",
    "RED" = "Red/Auburn",
    "SDY" = "Sandy",
    "WHI" = "White"
  )
  
  result <- hair_colors[code]
  return(ifelse(is.na(result), code, result))
}

#' Convert parsed AAMVA data to database-ready format
#'
#' @param parsed_data Result from parse_aamva_barcode()
#' @return Data frame ready for database insertion
prepare_for_database <- function(parsed_data) {
  
  if (!parsed_data$parsed) {
    stop("Cannot prepare unparsed data for database")
  }
  
  fields <- parsed_data$fields
  
  # Create data frame matching Patients table structure
  patient_data <- data.frame(
    license_number = as.character(fields$license_number),
    first_name = as.character(fields$first_name),
    last_name = as.character(fields$last_name),
    date_of_birth = as.character(fields$date_of_birth),
    address = as.character(fields$address_street),
    city = as.character(fields$address_city),
    state = as.character(fields$address_state),
    zip_code = as.character(fields$address_zip),
    phone = NA_character_,  # Not in barcode - will be entered manually
    email = NA_character_,  # Not in barcode - will be entered manually
    blood_type = NA_character_,  # Not in barcode - will be entered manually
    stringsAsFactors = FALSE
  )
  
  return(patient_data)
}

#' Format parsed data for display
#'
#' @param parsed_data Result from parse_aamva_barcode()
#' @return Formatted string for display
format_parsed_data <- function(parsed_data) {
  
  if (!parsed_data$parsed) {
    return(paste("Error:", parsed_data$error))
  }
  
  fields <- parsed_data$fields
  
  output <- paste0(
    "=== PARSED DRIVER'S LICENSE DATA ===\n\n",
    "PERSONAL INFORMATION:\n",
    "  Name: ", fields$first_name, " ", 
    ifelse(is.na(fields$middle_name), "", paste0(fields$middle_name, " ")),
    fields$last_name, "\n",
    "  Date of Birth: ", format(fields$date_of_birth, "%B %d, %Y"), "\n",
    "  Sex: ", fields$sex, "\n\n",
    
    "LICENSE INFORMATION:\n",
    "  License Number: ", fields$license_number, "\n",
    "  State: ", fields$address_state, "\n",
    "  Issue Date: ", format(fields$issue_date, "%B %d, %Y"), "\n",
    "  Expiration Date: ", format(fields$expiration_date, "%B %d, %Y"), "\n\n",
    
    "ADDRESS:\n",
    "  Street: ", fields$address_street, "\n",
    "  City: ", fields$address_city, "\n",
    "  State: ", fields$address_state, "\n",
    "  ZIP Code: ", fields$address_zip, "\n\n",
    
    "PHYSICAL DESCRIPTION:\n",
    "  Height: ", fields$height_inches, " inches\n",
    "  Weight: ", fields$weight_lbs, " lbs\n",
    "  Eye Color: ", fields$eye_color, "\n",
    "  Hair Color: ", fields$hair_color, "\n"
  )
  
  return(output)
}

# =====================================================
# TESTING FUNCTION
# =====================================================

#' Test the parser with sample data
test_parser <- function() {
  
  # Your actual Texas license scan
  test_data <- "ANSI 636015090002DL00410280ZT03210007DLDCACMDCBNONEDCDNONEDBA05252029DCSBARBERDDENDACBRYANDDFNDADEDWARDDDGNDBD06132025DBB05251977DBC1DAYHAZDAU072 inDAG2802 LAKESIDE LNDAICARROLLTONDAJTXDAK75006-4725DAQ10896644DCF20629580167103805092DCGUSADAZBRODCK10032767923DCLWDDAFDDB07162021DAW180DDK1ZTZTAN"
  
  cat("Testing AAMVA Parser with Texas Driver's License...\n\n")
  
  # Parse the data
  result <- parse_aamva_barcode(test_data)
  
  if (result$parsed) {
    cat(format_parsed_data(result))
    cat("\n\n=== DATABASE-READY FORMAT ===\n")
    db_data <- prepare_for_database(result)
    print(db_data)
    cat("\n✅ Parser test successful!\n")
  } else {
    cat("❌ Parser test failed:\n")
    cat(result$error, "\n")
  }
  
  return(result)
}

# Run test if this script is executed directly
if (!interactive()) {
  test_parser()
}

File: Python_Streamlit/aamva_parser.py
"""
AAMVA Driver's License Barcode Parser - Python Version
License to Live: MIAS
Parses AAMVA standard 2D barcodes from US driver's licenses
"""

import re
from datetime import datetime
from typing import Dict, Optional, Tuple

class AAMVAParser:
    """Parser for AAMVA standard driver's license barcodes"""
    
    # Field code mappings
    FIELD_CODES = {
        'DAC': 'first_name',
        'DAD': 'middle_name',
        'DCS': 'last_name',
        'DAQ': 'license_number',
        'DBB': 'date_of_birth',
        'DBD': 'issue_date',
        'DBA': 'expiration_date',
        'DAG': 'address_street',
        'DAI': 'address_city',
        'DAJ': 'address_state',
        'DAK': 'address_zip',
        'DBC': 'sex',
        'DAU': 'height_inches',
        'DAW': 'weight_lbs',
        'DAY': 'eye_color',
        'DAZ': 'hair_color',
        'DCG': 'country',
        'DCF': 'document_discriminator'
    }
    
    # Decode mappings
    SEX_MAP = {'1': 'Male', '2': 'Female', 'M': 'Male', 'F': 'Female'}
    
    EYE_COLOR_MAP = {
        'BLK': 'Black', 'BLU': 'Blue', 'BRO': 'Brown',
        'GRY': 'Gray', 'GRN': 'Green', 'HAZ': 'Hazel',
        'MAR': 'Maroon', 'PNK': 'Pink', 'DIC': 'Dichromatic'
    }
    
    HAIR_COLOR_MAP = {
        'BAL': 'Bald', 'BLK': 'Black', 'BLN': 'Blond',
        'BRO': 'Brown', 'GRY': 'Gray', 'RED': 'Red/Auburn',
        'SDY': 'Sandy', 'WHI': 'White'
    }
    
    def __init__(self):
        self.raw_data = None
        self.parsed = False
        self.error = None
        self.fields = {}
    
    def parse(self, barcode_data: str) -> Tuple[bool, Dict, Optional[str]]:
        """
        Parse AAMVA barcode data
        
        Args:
            barcode_data: Raw string from barcode scanner
            
        Returns:
            Tuple of (success, fields_dict, error_message)
        """
        self.raw_data = barcode_data
        self.parsed = False
        self.error = None
        self.fields = {}
        
        try:
            # Clean Eyoyo scanner output
            # Remove leading @ symbol
            cleaned_data = re.sub(r'^@\s*', '', barcode_data)
            
            # DON'T remove line breaks - use them to identify field boundaries!
            # Each AAMVA field starts on a new line
            
            # Remove leading/trailing whitespace from each line but keep line structure
            lines = cleaned_data.strip().split('\n')
            cleaned_lines = [line.strip() for line in lines if line.strip()]
            cleaned_data = '\n'.join(cleaned_lines)
            
            # Verify it starts with ANSI (AAMVA standard)
            if not cleaned_data.startswith('ANSI'):
                raise ValueError("Invalid barcode format - must start with ANSI")
            
            # Extract all fields
            for code, field_name in self.FIELD_CODES.items():
                value = self._extract_field(cleaned_data, code)
                
                # Special processing for certain fields
                if field_name in ['date_of_birth', 'issue_date', 'expiration_date']:
                    value = self._parse_date(value)
                elif field_name == 'sex':
                    value = self._decode_sex(value)
                elif field_name == 'eye_color':
                    value = self._decode_eye_color(value)
                elif field_name == 'hair_color':
                    value = self._decode_hair_color(value)
                
                self.fields[field_name] = value
            
            self.parsed = True
            return True, self.fields, None
            
        except Exception as e:
            self.error = f"Parse error: {str(e)}"
            self.parsed = False
            return False, {}, self.error
    
    def _extract_field(self, data: str, code: str) -> Optional[str]:
        """Extract a specific field from AAMVA data"""
        # Pattern: Find the field code at the start of a line, capture everything after it on that line
        # Each field is on its own line in the original format
        pattern = rf'^{code}(.*)$'
        match = re.search(pattern, data, re.MULTILINE)
        
        if match:
            value = match.group(1).strip()
            return value if value else None
        return None
    
    def _parse_date(self, date_str: Optional[str]) -> Optional[str]:
        """Parse AAMVA date format (MMDDYYYY) to YYYY-MM-DD"""
        if not date_str or len(date_str) != 8:
            return None
        
        try:
            month = date_str[0:2]
            day = date_str[2:4]
            year = date_str[4:8]
            
            # Validate date
            date_obj = datetime.strptime(f"{year}-{month}-{day}", "%Y-%m-%d")
            return date_obj.strftime("%Y-%m-%d")
        except ValueError:
            return None
    
    def _decode_sex(self, code: Optional[str]) -> Optional[str]:
        """Decode sex code"""
        if not code:
            return None
        return self.SEX_MAP.get(code.upper(), code)
    
    def _decode_eye_color(self, code: Optional[str]) -> Optional[str]:
        """Decode eye color code"""
        if not code:
            return None
        return self.EYE_COLOR_MAP.get(code.upper(), code)
    
    def _decode_hair_color(self, code: Optional[str]) -> Optional[str]:
        """Decode hair color code"""
        if not code:
            return None
        return self.HAIR_COLOR_MAP.get(code.upper(), code)
    
    def prepare_for_database(self) -> Dict:
        """Prepare parsed data for database insertion"""
        if not self.parsed:
            raise ValueError("Data not parsed yet")
        
        return {
            'license_number': self.fields.get('license_number'),
            'first_name': self.fields.get('first_name'),
            'last_name': self.fields.get('last_name'),
            'date_of_birth': self.fields.get('date_of_birth'),
            'address': self.fields.get('address_street'),
            'city': self.fields.get('address_city'),
            'state': self.fields.get('address_state'),
            'zip_code': self.fields.get('address_zip'),
            'phone': None,  # Not in barcode
            'email': None,  # Not in barcode
            'blood_type': None  # Not in barcode
        }
    
    def format_display(self) -> str:
        """Format parsed data for display"""
        if not self.parsed:
            return f"Error: {self.error}"
        
        output = "=== PARSED DRIVER'S LICENSE DATA ===\n\n"
        
        # Personal Information
        output += "PERSONAL INFORMATION:\n"
        name_parts = [self.fields.get('first_name')]
        if self.fields.get('middle_name'):
            name_parts.append(self.fields.get('middle_name'))
        name_parts.append(self.fields.get('last_name'))
        output += f"  Name: {' '.join(filter(None, name_parts))}\n"
        
        if self.fields.get('date_of_birth'):
            dob = datetime.strptime(self.fields['date_of_birth'], "%Y-%m-%d")
            output += f"  Date of Birth: {dob.strftime('%B %d, %Y')}\n"
        
        if self.fields.get('sex'):
            output += f"  Sex: {self.fields['sex']}\n"
        
        # License Information
        output += "\nLICENSE INFORMATION:\n"
        output += f"  License Number: {self.fields.get('license_number')}\n"
        output += f"  State: {self.fields.get('address_state')}\n"
        
        if self.fields.get('issue_date'):
            issue = datetime.strptime(self.fields['issue_date'], "%Y-%m-%d")
            output += f"  Issue Date: {issue.strftime('%B %d, %Y')}\n"
        
        if self.fields.get('expiration_date'):
            exp = datetime.strptime(self.fields['expiration_date'], "%Y-%m-%d")
            output += f"  Expiration Date: {exp.strftime('%B %d, %Y')}\n"
        
        # Address
        output += "\nADDRESS:\n"
        output += f"  Street: {self.fields.get('address_street')}\n"
        output += f"  City: {self.fields.get('address_city')}\n"
        output += f"  State: {self.fields.get('address_state')}\n"
        output += f"  ZIP Code: {self.fields.get('address_zip')}\n"
        
        # Physical Description
        output += "\nPHYSICAL DESCRIPTION:\n"
        if self.fields.get('height_inches'):
            output += f"  Height: {self.fields['height_inches']} inches\n"
        if self.fields.get('weight_lbs'):
            output += f"  Weight: {self.fields['weight_lbs']} lbs\n"
        if self.fields.get('eye_color'):
            output += f"  Eye Color: {self.fields['eye_color']}\n"
        if self.fields.get('hair_color'):
            output += f"  Hair Color: {self.fields['hair_color']}\n"
        
        return output


def test_parser():
    """Test the parser with sample Texas license data"""
    # Sample barcode from Eyoyo scanner (with @ and line breaks)
    test_data = """@
ANSI 636015090002DL00410280ZT03210007DLDCACM
DCBNONE
DCDNONE
DBA05252029
DCSBARBER
DDEN
DACBRYAN
DDFN
DADEDWARD
DDGN
DBD06132025
DBB05251977
DBC1
DAYHAZ
DAU072 in
DAG2802 LAKESIDE LN
DAICARROLLTON
DAJTX
DAK75006-4725
DAQ10896644
DCF20629580167103805092
DCGUSA
DAZBRO
DCK10032767923
DCLW
DDAF
DDB07162021
DAW180
DDK1
ZTZTAN"""
    
    print("Testing AAMVA Parser with Texas Driver's License...")
    print("(includes @ symbol and line breaks)\n")
    
    parser = AAMVAParser()
    success, fields, error = parser.parse(test_data)
    
    if success:
        print(parser.format_display())
        print("\n=== DATABASE-READY FORMAT ===")
        db_data = parser.prepare_for_database()
        for key, value in db_data.items():
            print(f"{key}: {value}")
        print("\n✅ Parser test successful!")
    else:
        print(f"❌ Parser test failed: {error}")
    
    return success


if __name__ == "__main__":
    # Run test if script is executed directly
    test_parser()
